Backend: Upgrade GORM and add PostgreSQL support

keif888 opened the pull request:
### Synopsis

This pull is branched from #4560 Replace GORM with GORM2.
It addresses #47 Config: Add support for PostgreSQL
It adds PostgreSQL as a supported DBMS for deploying PhotoPrism.

There is no intended functionality change provided by this pull.
There are a couple of new tests added, to ensure that case sensitivity in PostgreSQL is not negatively impacting the functionality of PhotoPrism.

### Status

All unit tests are passing.

### ToDo
- [X] Find out how to get psql into docker without having to run the command manually every time I restart the docker container
- [X] Create performance data creation functions for PostgreSQL
- [X] Run performance tests
- [X] Additional testing around case sensitivity
- [X] SQL improvements

Acceptance Criteria:

- [X] Features and enhancements must be fully implemented so that they can be released at any time without additional work
- [X] Automated unit and/or acceptance tests are mandatory to ensure the changes work as expected and to reduce repetitive manual work
- [X] Frontend components must be responsive to work and look properly on phones, tablets, and desktop computers; you must have tested them on all major browsers and different devices
- [X] Documentation and translation updates should be provided if needed
- [X] In case you submit database-related changes, they must be tested and compatible with SQLite 3 and MariaDB 10.5.12+

<!--

Since reviewing, testing and finally merging pull requests requires significant resources on our side, this can take several months if it's not just a small fix, especially if extensive testing is required to prevent bugs from getting into our stable version.

We thank you for your patience! :)

-->

lastzero commented:
@keif888 My understanding is that this PR should be good enough for frontend acceptance testing, i.e. there are no known issues left? Since the last batch of UI/UX enhancements [was released last week](https://github.com/photoprism/photoprism/releases/tag/250321-57590c48b), I'd now be able to build a testable Docker image from it, so our community can help test your changes :)

keif888 commented:
## Status
SQLite - all unit tests passing
PostgreSQL - 2 failing unit tests
- Sessions/SortByLoginAt
- Config_DatabaseVersion

MariaDB - no tests run yet

Once the unit tests pass on all three DBMS' this pull will be ready for making a testable docker image.

keif888 commented:
## Status
SQLite - all unit tests passing
PostgreSQL - all unit tests passing
MariaDB - all unit tests passing

Ready for extended user acceptance testing.

The previous failures causes were:
- Sessions/SortByLoginAt - PostgreSQL and MariaDB do not support nano seconds, which SQLite does
- Config_DatabaseVersion - Didn't have PostgreSQL support

lastzero commented:
@keif888 Thank you very much! We really appreciate your work, and so I am happy to report that your changes, along with my commits, have been merged into a new branch in our repository for further testing:

- https://github.com/photoprism/photoprism/commits/upgrade/gorm2-psql

The good news is that the code seems to compile and work, as I was able to start the development build and browse the pictures in my test library. **That's awesome!** 🥳

Unfortunately, I also saw a lot of errors in the unit tests. This could be because I merged/implemented some other pull requests before merging your changes to the new branch (which is based on develop), so e.g. the [background update of counts and covers](https://github.com/photoprism/photoprism/pull/4323) might be causing this?

Also, I ran "make dep-list" before the tests and then [updated the outdated packages](https://github.com/photoprism/photoprism/commit/2c7290868737ac490f9e30a9d1ea96c0a11dcdfa#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6), specifically:

- `gorm.io/driver/postgres` from v1.5.9 to v1.5.11
- `gorm.io/driver/sqlite` from v1.5.6 to v1.5.7

More likely, though, there are still some problems with the database schema(s) or my test environment, since one of the errors is "Unable to create FileWithoutPhoto.mp4 as it doesn't have required foreign key components". I've sent you my complete test output in the chat so you can check it. :white_check_mark: 

If the unit tests now require a running PostgreSQL instance, that might also be a problem, since I didn't update my containers before testing... I wouldn't mind adding a pre-configured `postgres` service based on the latest stable release to our [main compose.yaml file](https://github.com/photoprism/photoprism/blob/develop/compose.yaml) so that database testing doesn't require running additional commands?

We could also pre-install the `psql` command in the [Docker image](https://github.com/photoprism/photoprism/blob/develop/docker/develop/oracular/Dockerfile), if it isn't very large and doesn't install too many other dependencies (which I don't expect).

keif888 committed:
Merge remote-tracking branch 'photoprism/develop' into gorm2

keif888 committed:
Migration:  change sqlite3 to sqlite

keif888 committed:
Merge remote-tracking branch 'keif888/gorm2' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'photoprism/develop' into gorm2

keif888 committed:
Merge remote-tracking branch 'keif888/gorm2' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'photoprism/develop' into gorm2

keif888 committed:
Merge remote-tracking branch 'keif888/gorm2' into PostgreSQL

keif888 committed:
photo: correct preload of labels

keif888 committed:
Tests: Detect failed preload of labels

keif888 committed:
git:  add .vscode to ignore

keif888 committed:
CLI: improve detection of populated databases

keif888 committed:
Tests: Gorm V1 yaml files

keif888 committed:
Tests: Check more than just DeletedAt yaml errors

keif888 committed:
yaml: handle DeletedAt and other yaml errors

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
refactor: correct preload to exclude all photo fields

keif888 committed:
refactor: move queries from entity.

keif888 committed:
refactor: ensure that start from empty database will restore from yaml files without errors

keif888 committed:
tests: unit test that restore and index work with MediaRestoring

keif888 committed:
make: Add alldbms

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
refactor: make fmt-go

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Gorm: Fix missing & on Delete statements

keif888 committed:
Docker: Update to match compose.yml

keif888 committed:
Merge: Add missed index from merge

keif888 committed:
Tests: Update expected for Albums due new fixtures

keif888 committed:
Fixture: Renumber cows to 35 due duplicated id

keif888 committed:
Tests: make Moment tests survive changes to fixtures

keif888 committed:
Tests: update counts due to fix of album fixture duplicate id

keif888 committed:
Tests: Fix reuse of ID causing test failure

keif888 committed:
Gorm: Fix missing & on Model and Delete statements

keif888 committed:
PostgreSQL: Handle different ordering of NULL results

keif888 committed:
Migrations: Add Post step for post create/update processing

keif888 committed:
PostgreSQL: add ability to adjust order by for columns needing case insensitive order

keif888 committed:
PostgreSQL: add caseinsensitive collation and update tests

keif888 committed:
PostgreSQL: use ILIKE for case insensitive like processing

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
PostgreSQL: remove not required lower

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: correct assert.Equal order of operators

keif888 committed:
Build: update postgresql yaml file and adjust make names

keif888 committed:
Build: Shorten PostgreSQL to psql in make names

keif888 committed:
Build: Use postgres for consistency on all make commands

lastzero committed:
Build: Add "docker-postgres" target to Makefile

Signed-off-by: Michael Mayer <michael@photoprism.app>

lastzero commented:
Images that include these changes are now available for testing on Docker Hub:
- https://hub.docker.com/r/photoprism/photoprism/tags?name=postgres

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Docker: Update alldbms to match base compose.yaml

keif888 committed:
PostgreSQL: remove ToDo where it has been done.

keif888 committed:
CLI: Update usage descriptions for Boolean command flags

keif888 committed:
Tests: Make test TestConfig_PortalPath parallel compatible

keif888 committed:
Tests: remove ToDo that is done

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Setup: Update Ollama service examples in compose.alldbms.yaml file

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Backend: Refactor fmt function calls to use const string

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge: Fix merge errors

keif888 committed:
Docker: Update to match compose.yaml

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Backend: gorm2 update

keif888 committed:
Backend: correct merge error

keif888 committed:
Backend: add dbmutex for registry tests

PSandro commented:
Hi, great work!
I got a sql dialect warning when PhotoPrism was trying to merge two photos. It seems the code path for postgres does not exist for this case.
```
photoprism[2252]: time="2025-09-21T21:46:50+02:00" level=warning msg="sql: unsupported dialect postgres"
```
<img width="397" height="182" alt="image" src="https://github.com/user-attachments/assets/88e9e5fc-c587-4001-b1e7-4f68cf0c8932" />


https://github.com/keif888/photoprism/blob/f948cd4efb134013766ca467bf22fdba0bbe7c63/internal/entity/photo_merge.go#L120

keif888 committed:
Backend: fix missing postgres for photo_merge

keif888 committed:
Backend: add postgre and sqlite for UpdateAlbumDates

keif888 committed:
Backend: add postgres and sqlite for UpdateFolderDates

keif888 committed:
Config: close subscriptions as part of Shutdown

keif888 committed:
Tests: ensure cluster tests use DBMutex

keif888 committed:
DBMutex: detect case where test has Fatal, and left mutex in place

keif888 committed:
Config: Fix database close not working

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Make: apply reset-mariadb-all to postgres

keif888 committed:
Backend: apply changes for updated fs.Copy

keif888 committed:
Test: update test to support dynamic pathing

keif888 committed:
Backend: only open a database connection if it is closed

keif888 committed:
Test: fail test if it's not a recognised dbms

keif888 commented:
> Hi, great work! I got a sql dialect warning when PhotoPrism was trying to merge two photos. It seems the code path for postgres does not exist for this case.
> 
> ```
> photoprism[2252]: time="2025-09-21T21:46:50+02:00" level=warning msg="sql: unsupported dialect postgres"
> ```
Thanks for letting me know.
This has been fixed with commit [afe8fb3](https://github.com/photoprism/photoprism/pull/4831/commits/afe8fb33817290721162b4b860c177230225e156).
I rechecked all the switch statements and found 2 others that didn't have Postgres or SQLite.
I was sure I had done that before, but somehow I missed some.
This fix will be in the next Postgres push to Docker, when @lastzero gets time.

PSandro commented:
thanks, that seems to work perfectly now.
I'm now testing with your postgresql branch. (currently [45ee265](https://github.com/keif888/photoprism/commit/45ee265762c256763eb4419b5c59c78863dce129))
While trying to delete the Library->Errors logs, I found out that GORM would not allow batch delete without a where condition

`internal/entity/query/errors.go:30 WHERE conditions required [0.954ms] [rows:0] DELETE FROM "errors"`
https://github.com/keif888/photoprism/blob/45ee265762c256763eb4419b5c59c78863dce129/internal/entity/query/errors.go#L30

https://gorm.io/docs/delete.html#Block-Global-Delete :
_If you perform a batch delete without any conditions, GORM WON’T run it, and will return ErrMissingWhereClause error_

keif888 committed:
Query: add missing where clause to errors mass deletion

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Config: apply Dsn to DSN where not in merge

keif888 committed:
Tests: make cluster tests work when using different config dbms, and remove artifacts created in tests

keif888 committed:
Error: improve error subscription shutdown

keif888 commented:
> thanks, that seems to work perfectly now. I'm now testing with your postgresql branch. (currently [45ee265](https://github.com/keif888/photoprism/commit/45ee265762c256763eb4419b5c59c78863dce129)) While trying to delete the Library->Errors logs, I found out that GORM would not allow batch delete without a where condition
> 
> `internal/entity/query/errors.go:30 WHERE conditions required [0.954ms] [rows:0] DELETE FROM "errors"` https://github.com/keif888/photoprism/blob/45ee265762c256763eb4419b5c59c78863dce129/internal/entity/query/errors.go#L30
> 
> https://gorm.io/docs/delete.html#Block-Global-Delete : _If you perform a batch delete without any conditions, GORM WON’T run it, and will return ErrMissingWhereClause error_

Yeah, I knew about that, as I had coded delete statements for the test data refresh.
Sadly there was no test of that function, so it's failure hadn't been detected.
I have pushed a fix for that, and a test to confirm that it works.
I'm wading through the test coverage report to find what other database calls are not covered by tests.
There are 745 covered files, of which 184 are in the entity folders, and likely to have SQL that I need to check, and those have 79.5% code coverage.

keif888 committed:
Entity: Remove test database drop capability

keif888 committed:
Tests: add more coverage of sql

keif888 committed:
Config: fix sql execution in ImportSQL

keif888 committed:
Entity: add ILIKE for postgres and fixture update

keif888 committed:
Query: fix RemovePeopleAndFaces (delete without where and lack of no record found handling), and Sessions (ILIKE for postgres)

keif888 committed:
Query: revert ILIKE on auth_provider which is binary

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge: fix merge failures

keif888 committed:
Tests: handle auth change and fixture change

keif888 committed:
Entity: fix Drop and add migrations of tables covered by Once

keif888 committed:
Commands: improve non interactive handling

keif888 committed:
Tests: improve commands tests database handling, add more tests, disable hiding help in tests, fix remove use of DatabaseDSN for filenames, ensure migrations and versions are handled for tests and add busy_timeout for sqlite

keif888 committed:
Tests: fix wrong db in use after JWT tests

keif888 committed:
Tests: enable unstack StatusOK test

keif888 committed:
Tests: validate success

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: correct to int64 for db count and update number of fixtures

keif888 committed:
Entity: DeletedAt update in new fixture records for Gorm2

keif888 committed:
Tests: update benchmarks to use b.Loop

keif888 committed:
Backend: reinstate Values alias for map[string]interface{}

keif888 committed:
Backend: ensure that & is used for calls to Gorm

keif888 committed:
Tests: add logging message where fk violations are expected

keif888 committed:
Tests: improve dbmutex for sqlite and no dsn so that it uses file in storage/testdata

keif888 committed:
Tests: increase the permissable wait time in dbmutex

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Backend: address gorm2 migration changes

keif888 committed:
Backend: fix sqlite negative test and add postgres case

keif888 committed:
Tests: use PhotoPrismTestToDriverDsn

keif888 committed:
Tests: allow up to 8 packages to be tested simultaneously by creating multiple databases and locking a database to a package.  Move DSN to functions.  Move mariadb sql scripts to sub folder for consistency.

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge: handle changes resulting from merge

keif888 committed:
Merge: gorm2 required changes

keif888 committed:
Tests: change Fatal to Log if DropCredentials doesn't throw a User Drop error.  Add cleanup of the node registration

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Backend: move internal/functions to pkg

keif888 committed:
Backend: update testextras

keif888 committed:
Tests: Deleted records cause test failure due to selection difference in PostgreSQL

keif888 committed:
Compose: add postgres to default compose file

keif888 committed:
Query: add Postgres statements

keif888 committed:
Query: correct 0 to FALSE for postgres

keif888 committed:
Tests: remove db errors validation as no Db is setup in TestMain

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge: correct gorm statements

keif888 committed:
Compose: apply stop_grace_period to postgres

keif888 committed:
Compose: make the postgres compose file like the mariadb compose file

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Config: update to use NewMinimalTestConfig correctly

keif888 committed:
Cluster: update to use new casing of names in json.  Improve cleanup of tests

keif888 committed:
Tests: fix config test  for execution with DatabaseDSN populated

keif888 committed:
go: mod and sum auto updates

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: set maximum db parallel tests to 1 until TestImprovements is merged in

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
dsn: refactor PhotoPrismTestToDriverDsn to PhotoPrismTestToDriverDSN

keif888 committed:
Config: refactor Dsn to DSN

keif888 committed:
Entity: refactor Dsn to DSN

keif888 committed:
Testextras: Refactor PhotoPrismTestToDriverDsn to PhotoPrismTestToDriverDSN, ensure that target directory exists for Mutex DB and SQLite

keif888 committed:
Commands: remove commented log statements with Dsn

keif888 committed:
Make: remove photoprism.log on acceptance test start, handle DatabaseDsn or DatabaseDSN  for mariadb and postgres, split acceptance tests into separate lines to allow --ignore-errors to be used when there are known issues in acceptance tests

keif888 committed:
Tests: only check for db errors when using a db config

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: add postgres test

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge: fix issues due to merge (NewDSN moved to dsn, RoleInstance to RoleApp)

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Config: post merge issues

keif888 committed:
Dsn: Handle foreign keys, Postgres and merge issues

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: add missing showLogs to console.trace

keif888 committed:
Migration: fix failure to detect lower case data types in SQLite conversion

keif888 committed:
Dsn: improve ToString so that Params are alway used, add ForPSQL for DSN to support psql, and tests

keif888 committed:
Backup: replace hard coded dsn format with dsn.ToString

keif888 committed:
Tests: replace hard coded dsn's with dsn.ToString or dsn.ForPSQL

keif888 committed:
Config: update flag usage

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: add postgres dbdialect test

keif888 committed:
Merge tag '251130-b3068414c' into PostgreSQL

keif888 committed:
GormV2: Fixes required by Gorm V2

keif888 committed:
Clean: improve regex for Postgres and apply lint fixes

keif888 committed:
Tests: address changes from merge, fixture updates,

keif888 committed:
Migration: correct map to entity syntax

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Entity: Add reset of sequences/identity values for all tables with numeric ID

keif888 committed:
Tests: refactor restorePhoto to prevent TakenAt fields doubling up and breaking Postgres

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
make: ensure that migration databases are in place/refreshed for make test-go

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Makefile: Update docker-postgres to questing and include arm64

keif888 committed:
Compose: Update alldbms with latest from base compose file

keif888 committed:
Makefile: ensure storage/testdata exists for sqlite reset

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
tests: correct wrong file name in remove

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Makefile: add warnings about the resets that do the photoprism database

keif888 committed:
Config: use dsn to format DatabaseDSN

keif888 committed:
DSN: Add handling of Postgres URL encoding #5443

keif888 committed:
Test: Improve DSN related tests and add handling for special characters testing for Postgres

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Makefile: correct target reset-sqlite-unit

keif888 committed:
Scripts: Add portal database as per MariaDB

keif888 committed:
Tests: handle NewMinimalTestConfig difference

keif888 committed:
Portal: switch to GormV2

keif888 committed:
Tests: apply refactoring as per go-revive to remove redefinition of error and others.

keif888 committed:
Tests: remove file that wasn't testing but only logging (no valid tests in file)

keif888 committed:
Tests: refactor to remove redefinition of float32 and float64

keif888 committed:
Tests: refactor to ensure tests are asserting or requiring, don't use _ in names, and unused  setting

keif888 committed:
Tests: Validate 160 chars into Slug columns

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Backend: Refactor Gorm1 to Gorm 2

keif888 committed:
Tests: Refactor for changed NewMinimalTestConfig

keif888 committed:
Tests: Refactor away from SQLite database for error detections

keif888 committed:
Entity: Correct use of 0 instead of False which breaks Postgres.

keif888 committed:
Generate: Add Postgres to test data database generator

keif888 committed:
Tests: correct SyncSubject call back to false

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Compose: remove alldbms (now in compose.yaml), change host port for postgres, remove postgres:17-alpine due 18-alpine from merge

keif888 committed:
Makefile: add start's for each dbms

keif888 committed:
Tests: fix test so random chance doesn't break it

keif888 committed:
Postgres: correct query on refreshMonthAlbumCover

keif888 committed:
Covers: improve debug output, and improve tests so they manage test data to ensure they run

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Library: remove gorm.io due to reversion of .ppignore

keif888 committed:
Run `go fix ./...` and keep mechanical modernization updates.

- Replace `interface{}` with `any` in signatures and local types
- Apply formatter/style cleanups from go1.26 tooling
- Keep `omitempty` behavior-preserving simplifications suggested by fix
- No functional feature changes intended

Validation:
- make test-sqlite test-mariadb test-postgres  (Go 1.26.0)
- GOTOOLCHAIN=go1.24.10 make test-sqlite test-mariadb test-postgres

keif888 committed:
Run `go fix ./...` and keep mechanical modernization updates.

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge: correct merge issue with runewidth

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Cluster: Rename the "app" node role to "instance"

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
merge: address issues due to merge

keif888 committed:
merge: handle data type and function signature changes

keif888 committed:
Cluster: revert to devel branch test code

keif888 committed:
test: improve logging of expected sql messages

keif888 committed:
migrate: address gormv2 changes, and handle foreign keys

keif888 committed:
OIDC: add auth_id handling for GORMv2

keif888 committed:
CLI: fix user reset to include OIDC fix

keif888 committed:
migration: correct 20250819 to include if not exists

keif888 committed:
tests: save and reset config.Options so that test that change them don't break subsequent tests

keif888 committed:
migrate: correct the base mariadb table definition for auth_users

keif888 committed:
Migration: ensure that auth_users id column is the correct type

keif888 committed:
Migration: ensure that auth_users id column is the correct type

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
migrate: add transaction for sqlite table migration

keif888 committed:
commands: apply lint changes (gosec, spelling errors, comments, defer and os.exit)

keif888 committed:
tests: apply lint changes (gosec, spelling errors, comments, defer and os.exit)

keif888 committed:
migration: correct alter to unsigned for size column in marker table

keif888 committed:
benchmark: apply lint corrections

keif888 committed:
benchmark: improve generated database so markers are able to be used for faces update command

keif888 committed:
bemchmark: only generate a file if the appropriate test is run and it's not already there

keif888 committed:
generate: update to provide pseudo valid face embeddings

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
merge: g-runewidth update, NewMinimalTestConfig updates, and ExamplesPath changed to SamplesPath

keif888 committed:
Backend: Update dbms drivers

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Search: Remove group by handling for Postgres as group by has been removed from photos_geo

keif888 committed:
Go: Upgrade "jackc/pgx/v5" to v5.9.1

keif888 committed:
Go: mod tidy and fmt-go

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: ensure that refactored acceptance test target stops photoprism on database reset

keif888 committed:
Tests: cleanup acceptance-stop and remove photoprism.log if older than 3 days

keif888 committed:
Tests: address gorm2 count as int64

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Postgres: apply changes from mariadb-init.sql

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Add Dialects to dsn, and replace use of Driver with Dialect when checking the Gorm Dialector.Name().  Fix Merge issues with movement of consts to dsn.

keif888 committed:
make fmt-go applied

keif888 committed:
dsn: ensure postgres uses URI format by conversion of key value DSN's to URI format via ToString and Parse.  If a key value pair is passed in via individual parameters it will still work.

keif888 committed:
credentials: remove dsn.DriverTiDB

keif888 committed:
Tests: fix defer issue, and code refactors

keif888 committed:
Search: apply postgres required changes

keif888 committed:
Tests: refactor postgres test, update sqlite only test to skip other dbms, and add postgres only version of test

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Compose: make Postgres port available and starting from 4002

keif888 committed:
Tests: add video recording config, remove the d- on videoPath and videoOptions to enable.

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
MAkefile: add TC_USE_VIDEO environment varaible to trigger video recording in Testcafe.

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Compose: apply port environment config

keif888 committed:
go.mod fix incorrect runewidth from merge

keif888 committed:
Makefile: Apply dbms port from environment config

keif888 committed:
Dsn: utilise environment for default port numbers for testing

keif888 committed:
Tests: correct Driver

keif888 committed:
Makefile: increment postgres build to resolute

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge branch 'develop' into PostgreSQL

keif888 committed:
Tests: address merge/gorm/postgres issues

keif888 committed:
Config: remove Postgres not available

keif888 committed:
Search: gorm upgrades (TRUE/FALSE and int64)

keif888 committed:
Query: fix RemoveDuplicateMoments so it works with all DBMS'

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Backend: change type of NameSrc to match Gorm2 requirements.

keif888 committed:
Tests: update counter for new Session fixture

keif888 committed:
Backend: update jackc/pgx

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Database: Fix missing indexes, set MariaDB table creates to dynamic row format, correct defaults, enforce  albums table to dynamic row format for MariaDB

keif888 committed:
Database: fix defaults, remove defunct indexes

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: command output captures results so check for them

keif888 committed:
Fixtures: address gorm DeletedAt and increment counter due new lens fixture

keif888 committed:
Search: handle Postgres lack of case insensitive like

keif888 committed:
Tests: gorm 2 update and undo test result fully

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: ensure that null results in json don't break the acceptance tests

keif888 committed:
Entity: fix hard album delete not removing PhotosAlbums records, and correct Postgres time_index calculation not in 24h time

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: validate that time_index is 24h compatible

keif888 committed:
Tests: test that returned set from gorm v2 Scan are not nil (as per gorm v1)

keif888 committed:
Data: ensure that arrays in structs are populated as they were in Gorm v1.

keif888 committed:
Makefile: refactor the acceptance test targets to dbms generic targets

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: fix issue with int vs int64 for gorm2

keif888 committed:
Search: ensure that postgres nulls are handled correctly for users and sessions

keif888 committed:
Search: ensure that postgres nulls are handled correctly for users and sessions

keif888 committed:
Entity: address indexes on AlbumFilter and AlbumPath

keif888 committed:
tests: correct assertion so cleanup happens on no error.

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Tests: remove os.Exit from runTestMain so defers work.

keif888 committed:
Backend: Revert default changes

keif888 committed:
Tests: increase maximum concurrency of tests to 8 (from 1), fix tests failures due to competing mechanisms of test separation

keif888 committed:
Backend: Revert last 2 gorm default annotations

keif888 committed:
Tests: revert tests that had DBMS specific exceptions where all should work the same.

keif888 committed:
Search: ensure that DBMS case handling differences are ignored

keif888 committed:
Query: ensure that handling of limit < 0 is consistent as the DBMS' have different reactions

keif888 committed:
Tests: fix cache not being cleared issue between tests that reset the database (all DBMS).

keif888 committed:
Tests: Enable dynamic number of test databases based on cpu count, improve cleanup of databases

keif888 committed:
Tests: improve folder cleanup post tests, remove debug from database cleanup

keif888 committed:
Tests: revert use of dsn.PhotoPrismTestToFolderName to enable testing multiple DBMS at same time (not possible due to migrate commands) and replace with t.TempDir()

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Merge: Deal with post merge Gorm2, Postgres and function errors

keif888 committed:
Tests: enable per package databases on Postgres

keif888 committed:
Merge: Fix test failures related to merge including true/false instead of 1/0, increased fixtures, or spacing in messages

keif888 committed:
Tests: refactor the database test isolation capability from develop, replacing what I had produced.

keif888 committed:
Tests: remove unused testextras files

keif888 committed:
Query: correct use of 0/1 instead of FALSE/TRUE in sql statement

keif888 committed:
Tests: improve messaging about expected database error messages, fix missing directories for tests, add postgres test

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Makefile: update for Postgres

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

keif888 committed:
Makefile: add short form acceptance targets, and list in help

keif888 committed:
Tests: correct err checking, remove excess permission

keif888 committed:
Makefile: update mariadb/postgres targets for acceptance tests to address change in compose file

keif888 committed:
Compose: update mariadb and postgres dsn's so they have the superadmin account details

keif888 committed:
Merge remote-tracking branch 'origin/develop' into PostgreSQL

photoprismci commented:
A new test image has been built from the current tip of this branch (`714f81cda` — "Merge remote-tracking branch 'origin/develop' into PostgreSQL") and is now available on Docker Hub:

- `photoprism/photoprism:postgres` (rolling)
- `photoprism/photoprism:260731-postgres` (datestamped)

Both `linux/amd64` and `linux/arm64` are included in the manifest. You can pull it with:

```sh
docker pull photoprism/photoprism:postgres
```

Thanks for the continued progress on this — looking forward to hearing how it goes!

keif888 committed:
Tests: fix test suite being marked as passed when it failed by database cleanup

keif888 committed:
Internal: apply cleanups and merge misses

keif888 committed:
Tests: apply cleanups and merge misses

keif888 committed:
Convert: update copyright date

keif888 committed:
SQL: remove not required permission CREATEDB

keif888 committed:
Tests: remove no longer used parameter to match develop branch

lastzero commented:
Just a heads-up on the naming of the test databases:

We spent some time this week planning to rename the MariaDB unit test databases on `develop` because `reset-mariadb-acceptance` resets the per-package Go test databases, whereas TestCafe never touches MariaDB at all - the name of the target suggests the opposite of what it does. We then found that you had already done it here in a better way than we had planned: unit tests on `testdb` and TestCafe accessing the genuine `acceptance` databases on MariaDB and Postgres via the Makefile variables `MARIADB_ACPT_DSN` and `POSTGRES_ACPT_DSN`. The reset scripts are also split per driver. Yours is the better solution - keep it.

We are therefore not changing this on `develop`, i.e. no renaming and we won't drop the `acceptance` MariaDB user either. It isn't used on `develop` at the moment as every test DSN connects as `root:photoprism`, so it seemed redundant - but the `$(subst root:photoprism,acceptance:acceptance,...)` lines in this branch depend on it, so removing it would have broken your work.
